app/controllers/concerns/classified_listings_toolkit.rb

```diff
@@ -34,4 +36,170 @@ def bump_listing_success
+        cost: cost,
+      )
+
+      # save the listing
+      @classified_listing.bumped_at = Time.current
+      @classified_listing.published = true
+
+      # since we can't raise active record errors in this transaction
+      # due to the fact that we need to display them in the :new view,
+      # we manually rollback the transaction if there are validation errors
+      raise ActiveRecord::Rollback unless @classified_listing.save
+
+      successful_transaction = true
+    end
+
+    if successful_transaction
+      clear_listings_cache
+      @classified_listing.index!
+      process_successful_creation
+    else
+      @credits = current_user.credits.unspent
+      @classified_listing.cached_tag_list = listing_params[:tag_list]
+      @organizations = current_user.organizations
+      process_unsuccessful_creation
+    end
+  end
+
+  def update
+    authorize @classified_listing
+
+    cost = ClassifiedListing.cost_by_category(@classified_listing.category)
+
+    # NOTE: this should probably be split in three different actions: bump, unpublish, publish
+    return bump_listing(cost) if listing_params[:action] == "bump"
+
+    if listing_params[:action] == "unpublish"
+      unpublish_listing
+      process_after_unpublish
+      return
+    elsif listing_params[:action] == "publish"
+      unless @classified_listing.bumped_at?
+        first_publish(cost)
+        return
+      end
+
+      publish_listing
+    elsif listing_updatable?
+      update_listing_details
+    end
+
+    clear_listings_cache
+    process_after_update
+  end
+
+  def bump_listing(cost)
+    org = Organization.find_by(id: @classified_listing.organization_id)
+
+    available_org_credits = org.credits.unspent if org
+    available_user_credits = current_user.credits.unspent
+
+    if org && available_org_credits.size >= cost
+      charge_credits_before_bump(org, cost)
+    elsif available_user_credits.size >= cost
+      charge_credits_before_bump(current_user, cost)
+    else
+      process_no_credit_left && return
+    end
+  end
+
+  def charge_credits_before_bump(purchaser, cost)
+    ActiveRecord::Base.transaction do
+      Credits::Buyer.call(
+        purchaser: purchaser,
+        purchase: @classified_listing,
+        cost: cost,
+      )
+
+      raise ActiveRecord::Rollback unless bump_listing_success
+    end
+  end
+
+  def first_publish(cost)
+    available_author_credits = @classified_listing.author.credits.unspent
+    available_user_credits = []
+    if @classified_listing.author.is_a?(Organization)
+      available_user_credits = current_user.credits.unspent
+    end
+
+    if available_author_credits.size >= cost
+      create_listing(@classified_listing.author, cost)
+    elsif available_user_credits.size >= cost
+      create_listing(current_user, cost)
+    else
+      process_no_credit_left
+    end
+  end
+
+  def listing_updatable?
+    at_leat_one_param_present? && (bumped_in_last_24_hrs? || !@classified_listing.published)
```

Comment:
can you explain how you arrived to this logic? or point me to the conversation.

ps. `at_leat_one_param_present` as a typo, it should be `at_least_one_param_present`

Reply:
I have just decomposed the previous code and refactored the logic in many methods.